lightning-invoice/src/utils.rs

```diff
@@ -21,6 +21,98 @@ use core::convert::TryInto;
+/// load-balancing payments between multiple nodes.
+///
+/// `channels` parameter:
+/// * Contains a list of tuples of `(phantom_node_scid, real_node_pubkey,
+///   vec[real_channel_details])`, where `phantom_node_scid` is unique to each entry
+/// * `phantom_node_scid`s are retrieved from [`ChannelManager::get_phantom_scid`]
+/// * Each `vec[real_channel_details]` is retrieved from [`ChannelManager::list_usable_channels`]
+/// * Contains channel info for all nodes participating in the phantom invoice
+/// * It is fine to cache `channels` and reuse it across invoices, as long as the data is
+///   updated when a channel becomes disabled or closes
+/// * Note that including too many channels here could result in making the invoice too long for QR
+///   code scanning
+///
+/// `payment_hash` and `payment_secret` come from [`ChannelManager::create_inbound_payment`] or
+/// [`ChannelManager::create_inbound_payment_for_hash`]. These values can be retrieved from any
+/// participating node.
+///
+/// See [`KeysInterface::get_phantom_secret`] for more requirements on supporting phantom node
+/// payments.
+///
+/// [`KeysInterface::get_phantom_secret`]: lightning::chain::keysinterface::KeysInterface::get_phantom_secret
+pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
+	amt_msat: Option<u64>, description: String, payment_hash: PaymentHash, payment_secret:
+	PaymentSecret, channels: &[(u64, PublicKey, Vec<ChannelDetails>)], keys_manager: K, network: Currency
+) -> Result<Invoice, SignOrCreationError<()>> where K::Target: KeysInterface {
+	if channels.len() == 0 {
+		return Err(SignOrCreationError::CreationError(CreationError::MissingRouteHints))
+	}
+	let mut invoice = InvoiceBuilder::new(network)
+		.description(description)
+		.current_timestamp()
+		.payment_hash(Hash::from_slice(&payment_hash.0).unwrap())
+		.payment_secret(payment_secret)
+		.min_final_cltv_expiry(MIN_FINAL_CLTV_EXPIRY.into());
+	if let Some(amt) = amt_msat {
+		invoice = invoice.amount_milli_satoshis(amt);
+	}
+
+	for (phantom_scid, real_node_pubkey, chan_details_list) in channels {
+		for channel in chan_details_list {
+			let short_channel_id = match channel.short_channel_id {
+				Some(id) => id,
+				None => continue,
+			};
+			let forwarding_info = match &channel.counterparty.forwarding_info {
+				Some(info) => info.clone(),
+				None => continue,
+			};
+			invoice = invoice.private_route(RouteHint(vec![
+					RouteHintHop {
+						src_node_id: channel.counterparty.node_id,
+						short_channel_id,
+						fees: RoutingFees {
+							base_msat: forwarding_info.fee_base_msat,
+							proportional_millionths: forwarding_info.fee_proportional_millionths,
+						},
+						cltv_expiry_delta: forwarding_info.cltv_expiry_delta,
+						htlc_minimum_msat: None,
+						htlc_maximum_msat: None,
+					},
+					RouteHintHop {
+						src_node_id: *real_node_pubkey,
+						short_channel_id: *phantom_scid,
```

Comment:
I think the utility comment should be clearer that phantom invoice does not provide recipient anonymity to avoid misuses. Especially as the feature name "phantom" suggest some privacy advantages.

As the hop servicing as a callback is always introduced at a fixed offset, the phantom invoice receiver can deduce that the last element of each routing hints sequence is part of the phantom nodes set.

Reply:
Not really my place to comment, but in case anyone is interested: phantom comes from the same root as fantasy, so it sounds like imaginary invoice (aside from the privacy connotations noted above). Maybe meta invoice would be more fitting? (among other things, meta means "between")

Reply:
Hmm, would be surprised if users assumed privacy advantages based on just the name "phantom". Added a brief comment though.

> Maybe meta invoice would be more fitting? (among other things, meta means "between")

Thanks for the review. I'd be hesitant to think this rename is a big win given users so far seem to find the phantom concept to be a decently intuitive mental model, plus it'd be a significant change at this point..